AD Theme Direct WFS 2.0 Test - No Temporal values found

To the OGC Community,

The following is the error:
 
OGC TEAM Engine reported a failed test: No temporal values found in results: property is "http://inspire.ec.europa.eu/schemas/ad/4.0 [inspire.ec.europa.eu]":beginLifespanVersion. expected [false] but found [true]
 
This is the decoded request that the validator sends
 
http://agsinspire1.esri.com/arcgis/rest/services/ADTest/MapServer/exts/InspireFeatureDownload/service?service=WFS&version=2.0.0&request=GetFeature&count=10&typenames=ns63:ThoroughfareName&namespaces=xmlns(xml,http://www.w3.org/XML/1998/namespace),xmlns(ns63,http://inspire.ec.europa.eu/schemas/ad/4.0),xmlns(wfs,http://www.opengis.net/wfs/2.0)&filter=<fes:Filter xmlns:ns63="http://inspire.ec.europa.eu/schemas/ad/4.0 [inspire.ec.europa.eu]" xmlns:wfs="http://www.opengis.net/wfs/2.0 [opengis.net]" xmlns:fes="http://www.opengis.net/fes/2.0 [opengis.net]"><fes:During><fes:ValueReference xmlns:tns="http://inspire.ec.europa.eu/schemas/ad/4.0 [inspire.ec.europa.eu]">tns:beginLifespanVersion</fes:ValueReference><gml:TimePeriod xmlns:gml="http://www.opengis.net/gml/3.2 [opengis.net]" frame="http://www.iso.org/iso/iso8601 [iso.org]" gml:id="TP01"><gml:beginPosition>1988-09-05T10:00:00Z</gml:beginPosition><gml:endPosition>1988-09-05T12:00:00Z</gml:endPosition></gml:TimePeriod></fes:During></fes:Filter>
 
 
The existing beginLifespanVersion value in the database for the ThoroughfareName entry is: 9/5/1988 6:00:00 AM
If I adjust the request’s gml:beginPosition and gml:endPosition to the following highlighted below, by taking the “Z” off the time and adjusting the beginPosition time an hour before the current value, then I get the returned component.
 
 
http://agsinspire1.esri.com/arcgis/rest/services/ADTest/MapServer/exts/InspireFeatureDownload/service?service=WFS&version=2.0.0&request=GetFeature&count=10&typenames=ns63:ThoroughfareName&namespaces=xmlns(xml,http://www.w3.org/XML/1998/namespace),xmlns(ns63,http://inspire.ec.europa.eu/schemas/ad/4.0),xmlns(wfs,http://www.opengis.net/wfs/2.0)&filter=<fes:Filter xmlns:ns63="http://inspire.ec.europa.eu/schemas/ad/4.0 [inspire.ec.europa.eu]" xmlns:wfs="http://www.opengis.net/wfs/2.0 [opengis.net]" xmlns:fes="http://www.opengis.net/fes/2.0 [opengis.net]"><fes:During><fes:ValueReference xmlns:tns="http://inspire.ec.europa.eu/schemas/ad/4.0 [inspire.ec.europa.eu]">tns:beginLifespanVersion</fes:ValueReference><gml:TimePeriod xmlns:gml="http://www.opengis.net/gml/3.2 [opengis.net]" frame="http://www.iso.org/iso/iso8601 [iso.org]" gml:id="TP01"><gml:beginPosition>1988-09-05T05:00:00</gml:beginPosition><gml:endPosition>1988-09-05T12:00:00</gml:endPosition></gml:TimePeriod></fes:During></fes:Filter>
 

Our service returns datetimes without timezones. This means (as per XML Schema spec) the datetimes are in the servers local timezone. This timezone is unknown for external users, because it is not declared by the service. 

I am not aware that the WFS spec supports declaring the timezone by the service.

Also, it almost seems like the Validator is assuming the datetimes returned by the service are in the local timezone of the validator.

Can someone please assess this issue and give us guidance?

Thank you.

@bpross-52n I have a similar issue with [this WFS](https://geodienste.komm.one/ows/services/org.506.78427a5b-2f82-4311-ba7d-74eb7b8c10f4_wfs/org.506.e21f771b-0d06-4dbc-b246-60aa8d12ed52?SERVICE=WFS&REQUEST=GetCapabilities&VERSION=2.0.0).
![grafik](https://user-images.githubusercontent.com/34163838/161543879-05d0ad5a-d7fd-4572-801b-1a05fb7487fe.png)
There are two kinds of error messages:
1. No temporal values found in results: property is "http://inspire.ec.europa.eu/schemas/plu/4.0":validFrom.  --> there is a validFrom value in the dataset in the SpatialPlan object (the OfficialDoument objects do not need one - see [UML](https://inspire.ec.europa.eu/data-model/approved/r4618-ir/html/EARoot/EA2/EA3/EA10/EA1/EA4/EA8445.htm)) and it is valid according to the INSPIRE Validator (dataset and report: 
[validFrom.zip](https://github.com/opengeospatial/ets-wfs20/files/8409689/validFrom.zip) )
Could you please check why the error message is occurring?
2. java.lang.AssertionError: t1 (Period: 2012-04-04T00:00:00.000+0000/2012-04-04T23:59:59.999+0000) is not related to t2 (Period: 2012-04-04T16:19:59.999+0000/2012-04-05T00:59:59.998+0000): BEFORE. The actual temporal relation is OVERLAPS -> not sure what t1 and t2 are in this case and what would be invalid about them.

If you want me to open a separate issue on that, please let me know. But as one of the error messages is the same as in this issue, I thought it would make sense to add it here.


Thank you and all the best

Johanna

> @bpross-52n I have a similar issue with [this WFS](https://geodienste.komm.one/ows/services/org.506.78427a5b-2f82-4311-ba7d-74eb7b8c10f4_wfs/org.506.e21f771b-0d06-4dbc-b246-60aa8d12ed52?SERVICE=WFS&REQUEST=GetCapabilities&VERSION=2.0.0). ![grafik](https://user-images.githubusercontent.com/34163838/161543879-05d0ad5a-d7fd-4572-801b-1a05fb7487fe.png) There are two kinds of error messages:
> 
> 1. No temporal values found in results: property is "[http://inspire.ec.europa.eu/schemas/plu/4.0":validFrom](http://inspire.ec.europa.eu/schemas/plu/4.0%22:validFrom).  --> there is a validFrom value in the dataset in the SpatialPlan object (the OfficialDoument objects do not need one - see [UML](https://inspire.ec.europa.eu/data-model/approved/r4618-ir/html/EARoot/EA2/EA3/EA10/EA1/EA4/EA8445.htm)) and it is valid according to the INSPIRE Validator (dataset and report:
>    [validFrom.zip](https://github.com/opengeospatial/ets-wfs20/files/8409689/validFrom.zip) )
>    Could you please check why the error message is occurring?
> 2. java.lang.AssertionError: t1 (Period: 2012-04-04T00:00:00.000+0000/2012-04-04T23:59:59.999+0000) is not related to t2 (Period: 2012-04-04T16:19:59.999+0000/2012-04-05T00:59:59.998+0000): BEFORE. The actual temporal relation is OVERLAPS -> not sure what t1 and t2 are in this case and what would be invalid about them.
> 
> If you want me to open a separate issue on that, please let me know. But as one of the error messages is the same as in this issue, I thought it would make sense to add it here.
> 
> Thank you and all the best
> 
> Johanna

@JohannaOtt 
Regarding the first point: This is fixed on the beta instance of Teamengine.

Regarding the second point: This will be fixed by #241, see also here: [comment](https://github.com/opengeospatial/ets-wfs20/issues/226#issuecomment-1276195224)

--- a/pom.xml
+++ b/pom.xml
@@ -51,9 +51,9 @@
     <soapui.teamengine.user>ogctest</soapui.teamengine.user>
     <soapui.teamengine.password>ogctest</soapui.teamengine.password>
     <soapui.iut>https://cite.deegree.org/deegree-webservices-3.4.30/services/wfs200?service=WFS&amp;request=GetCapabilities</soapui.iut>
-    <soapui.tests.passed>132</soapui.tests.passed>
-    <soapui.tests.skipped>38</soapui.tests.skipped>
-    <soapui.tests.failed>3</soapui.tests.failed>
+    <soapui.tests.passed>120</soapui.tests.passed>
+    <soapui.tests.skipped>53</soapui.tests.skipped>
+    <soapui.tests.failed>0</soapui.tests.failed>
     <docker.teamengine.version>5.4.1</docker.teamengine.version>
   </properties>
 

--- a/src/main/java/org/opengis/cite/iso19142/basic/filter/temporal/AbstractTemporalTest.java
+++ b/src/main/java/org/opengis/cite/iso19142/basic/filter/temporal/AbstractTemporalTest.java
@@ -23,7 +23,16 @@ public abstract class AbstractTemporalTest extends QueryFilterFixture {
             throw new SkipException( "Feature type has no temporal properties: " + featureType );
         }
 
-        TemporalProperty temporalExtent = findTemporalExtent( featureType, temporalProperties );
+        TemporalProperty temporalExtent = null;
+        
+        try {
+        	temporalExtent = findTemporalExtent( featureType, temporalProperties );
+		} catch (Exception e) {
+            if(e instanceof SkipException) {
+            	throw e;
+            }
+		}
+        
         if ( temporalExtent == null )
             throw new SkipException(
                                      "Feature type + "
@@ -42,6 +51,9 @@ public abstract class AbstractTemporalTest extends QueryFilterFixture {
             } catch ( Exception e ) {
                 LOGR.warning( "Could not calculate the extent of the temporal property " + temporalProp
                               + " of the feature type " + featureType );
+                if(e instanceof SkipException) {
+                	throw e;
+                }
             }
         }
         return null;

--- a/src/main/java/org/opengis/cite/iso19142/basic/filter/temporal/TemporalQuery.java
+++ b/src/main/java/org/opengis/cite/iso19142/basic/filter/temporal/TemporalQuery.java
@@ -86,7 +86,6 @@ public class TemporalQuery {
      * @return A TemporalGeometricPrimitive instance (instant or period).
      */
     public static TemporalGeometricPrimitive parseTemporalValue(String value, XSTypeDefinition typeDefinition) {
-//    	value = value.replace("Z", "");
         if ( typeDefinition.getTypeCategory() != XSTypeDefinition.SIMPLE_TYPE
              && !( ( (XSComplexTypeDefinition) typeDefinition ).getContentType() == XSComplexTypeDefinition.CONTENTTYPE_SIMPLE ) ) {
             throw new IllegalArgumentException( "Not a simple type definition: " + typeDefinition.getName() );
@@ -104,7 +103,7 @@ public class TemporalQuery {
             DateTimeFormatter xsdDateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss[.SSS][XXX]");
             TemporalAccessor tm = xsdDateTimeFormatter.parseBest(value, ZonedDateTime::from, LocalDateTime::from);
             if (tm instanceof LocalDateTime) {
-            	throw new SkipException("Local timezones currently not supported.");
+            	throw new SkipException("{%s uses date values without timezone, which are currently not supported by this test suite.");
             }
             ZonedDateTime dateTime = (ZonedDateTime) tm;
             tmPrimitive = tmFactory.createInstant(new DefaultPosition(Date.from(dateTime.toInstant())));
@@ -112,7 +111,7 @@ public class TemporalQuery {
         case XSConstants.DATE_DT:
             ZoneOffset zone = DateTimeFormatter.ISO_DATE.parse(value, TemporalQueries.offset());
             if (null == zone) {
-                zone = ZonedDateTime.now().getOffset();
+            	throw new SkipException("{%s uses date values without timezone, which are currently not supported by this test suite.");
             }
             LocalDate date = LocalDate.parse(value, DateTimeFormatter.ISO_DATE);
             Temporal startOfDay = date.atStartOfDay(zone);

--- a/src/main/java/org/opengis/cite/iso19142/basic/filter/temporal/TemporalQuery.java
+++ b/src/main/java/org/opengis/cite/iso19142/basic/filter/temporal/TemporalQuery.java
@@ -28,6 +28,7 @@ import org.opengis.cite.iso19142.Namespaces;
 import org.opengis.cite.iso19142.util.WFSMessage;
 import org.opengis.temporal.TemporalFactory;
 import org.opengis.temporal.TemporalGeometricPrimitive;
+import org.testng.SkipException;
 import org.w3c.dom.Document;
 import org.w3c.dom.Node;
 
@@ -85,6 +86,7 @@ public class TemporalQuery {
      * @return A TemporalGeometricPrimitive instance (instant or period).
      */
     public static TemporalGeometricPrimitive parseTemporalValue(String value, XSTypeDefinition typeDefinition) {
+//    	value = value.replace("Z", "");
         if ( typeDefinition.getTypeCategory() != XSTypeDefinition.SIMPLE_TYPE
              && !( ( (XSComplexTypeDefinition) typeDefinition ).getContentType() == XSComplexTypeDefinition.CONTENTTYPE_SIMPLE ) ) {
             throw new IllegalArgumentException( "Not a simple type definition: " + typeDefinition.getName() );
@@ -102,8 +104,7 @@ public class TemporalQuery {
             DateTimeFormatter xsdDateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss[.SSS][XXX]");
             TemporalAccessor tm = xsdDateTimeFormatter.parseBest(value, ZonedDateTime::from, LocalDateTime::from);
             if (tm instanceof LocalDateTime) {
-                // set local time zone
-                tm = LocalDateTime.class.cast(tm).atZone(ZonedDateTime.now().getOffset());
+            	throw new SkipException("Local timezones currently not supported.");
             }
             ZonedDateTime dateTime = (ZonedDateTime) tm;
             tmPrimitive = tmFactory.createInstant(new DefaultPosition(Date.from(dateTime.toInstant())));

--- a/src/main/java/org/opengis/cite/iso19142/util/DataSampler.java
+++ b/src/main/java/org/opengis/cite/iso19142/util/DataSampler.java
@@ -54,6 +54,7 @@ import org.opengis.cite.iso19142.basic.filter.temporal.TemporalQuery;
 import org.opengis.geometry.Envelope;
 import org.opengis.temporal.Period;
 import org.opengis.temporal.TemporalGeometricPrimitive;
+import org.testng.SkipException;
 import org.w3c.dom.Document;
 import org.w3c.dom.Element;
 import org.w3c.dom.NodeList;
@@ -453,6 +454,9 @@ public class DataSampler {
                 tmSet.add(tVal);
             } catch (RuntimeException re) {
                 LOGR.log(Level.WARNING, re.getMessage());
+                if(re instanceof SkipException) {
+                	throw re;
+                }
                 continue;
             }
         }

--- a/src/main/java/org/opengis/cite/iso19142/util/DataSampler.java
+++ b/src/main/java/org/opengis/cite/iso19142/util/DataSampler.java
@@ -455,7 +455,9 @@ public class DataSampler {
             } catch (RuntimeException re) {
                 LOGR.log(Level.WARNING, re.getMessage());
                 if(re instanceof SkipException) {
-                	throw re;
+                	String message = re.getMessage();
+                	message = String.format(message, featureType.toString());
+                	throw new SkipException(message);
                 }
                 continue;
             }

--- a/src/test/java/org/opengis/cite/iso19142/basic/filter/temporal/VerifyTemporalQuery.java
+++ b/src/test/java/org/opengis/cite/iso19142/basic/filter/temporal/VerifyTemporalQuery.java
@@ -26,6 +26,7 @@ import org.opengis.cite.validation.XmlSchemaCompiler;
 import org.opengis.temporal.Instant;
 import org.opengis.temporal.Period;
 import org.opengis.temporal.TemporalGeometricPrimitive;
+import org.testng.SkipException;
 import org.w3c.dom.Document;
 import org.w3c.dom.Node;
 import org.xml.sax.SAXException;
@@ -58,10 +59,16 @@ public class VerifyTemporalQuery extends CommonTestFixture {
     @Test
     public void parseDate() {
         XSTypeDefinition typeDef = model.getTypeDefinition("date", XMLConstants.W3C_XML_SCHEMA_NS_URI);
-        TemporalGeometricPrimitive result = TemporalQuery.parseTemporalValue("2016-05-15", typeDef);
+        TemporalGeometricPrimitive result = TemporalQuery.parseTemporalValue("2016-05-15Z", typeDef);
         assertTrue("Expected result: " + Period.class.getName(), Period.class.isInstance(result));
         Period period = Period.class.cast(result);
         assertTrue("Expected duration: PT23H59M59S", period.length().toString().equals("PT23H59M59S"));
+    }
+
+    @Test(expected = SkipException.class)
+    public void parseDateWithoutTimezone() {
+        XSTypeDefinition typeDef = model.getTypeDefinition("date", XMLConstants.W3C_XML_SCHEMA_NS_URI);
+        TemporalQuery.parseTemporalValue("2016-05-15", typeDef);
     }
 
     @Test

